File: ruby/rotational-cipher/rotational_cipher.rb
module RotationalCipher
  CHARACTERS = %W{a b c d e f g h i j k l m n o p q r s t u v w x y z}

  def self.rotate(text, key)
    text.gsub(/\w/) { |chr| rotate_character(chr, key) }
  end

  def self.rotate_character(chr, key)
    CHARACTERS[(CHARACTERS.index(chr) + key) % CHARACTERS.length]
  end
end

File: javascript/rotational-cipher/rotational-cipher.js
const AVALIABLE_ROTATE_CHRS = /[a-zA-Z]/g;
const ASCII_START_CODE = 97;
const ALPHABET_LENGTH = 26;

export class RotationalCipher {
  static rotate(input, key) {
    return input.replace(AVALIABLE_ROTATE_CHRS, chr =>
      this.rotateChr(chr, key)
    );
  }

  static rotateChr(chr, key) {
    if (chr === chr.toUpperCase()) {
      const lowerChr = chr.toLowerCase();

      return this.rotateChr(lowerChr, key).toUpperCase();
    }

    const inputIndex = chr.charCodeAt(0) - ASCII_START_CODE;
    const outputIndex =
      ((inputIndex + key) % ALPHABET_LENGTH) + ASCII_START_CODE;

    return String.fromCharCode(outputIndex);
  }
}
